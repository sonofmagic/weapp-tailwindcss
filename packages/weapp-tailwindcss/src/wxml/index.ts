export { replaceWxml } from './shared'
export { generateCode, templateReplacer, customTemplateHandler, createTemplateHandler } from './utils'
