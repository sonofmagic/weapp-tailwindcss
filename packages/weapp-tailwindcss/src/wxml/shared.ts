import type { ITemplateHandlerOptions } from '@/types'
import { SimpleMappingChars2String, escape } from '@/escape'

export function replaceWxml(
  original: string,
  options: ITemplateHandlerOptions = {
    keepEOL: false,
    escapeMap: SimpleMappingChars2String,
  },
) {
  const { keepEOL, escapeMap, mangleContext } = options
  let res = original
  if (!keepEOL) {
    res = res
      // 去除无用换行符和空格
      // 不能全去掉，头条小程序变量绑定，实现方式依赖空格，你说坑不坑？
      .replaceAll(/[\n\r]+/g, '')
  }
  if (mangleContext) {
    res = mangleContext.wxmlHandler(res)
  }

  res = escape(res, {
    map: escapeMap,
  })

  return res
}
