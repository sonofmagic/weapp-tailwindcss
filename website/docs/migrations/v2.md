# 从 v2 迁移到 v3

v3 版本相比于 v2, 主要是删去一些过时的功能，配置项，同时会改变插件的默认值，使得整体插件变得更易用，更容易安装
假如你没有用到什么复杂自定义配置，那么完全可以平滑升级上来。

## 配置项改动

### 删除的配置项

- 删去 `replaceUniversalSelectorWith` 选项，使用 `cssSelectorReplacement.universal` 来代替，后者参数覆盖前者
- 删去 `minifiedJs` 选项，现在完全遵从用户的配置，用户压缩就压缩，反之亦然
- 删去 `jsEscapeStrategy` 选项，现在默认只有一种模式 `replace`/  不再提供 `regenerate` 模式
- 删去 `customReplaceDictionary` 的 `complex` 模式，只内置 `simple` 模式 (你如果还要 `complex` 模式 ,可以从 `@weapp-core/escape` 引入，再传入 `customReplaceDictionary` 配置项即可)
- `cssMatcher`/`htmlMatcher`/`jsMatcher`/ `mainCssChunkMatcher` / `wxsMatcher` 不再能够传入 `glob` 表达式(例如`**/*.html`)，现在都是传入一个方法: `(name: string) => boolean`。要兼容原先的 `glob` 表达式，你可以通过 `minimatch` 把 `glob` 表达式转化成正则来兼容原先的配置
- `cssPreflightRange` 只存在一种模式，为 `all`, 之前的 `view` 选项交给 `cssSelectorReplacement.universal` 进行托管

### 增加的配置项

- `rem2rpx` : 类型 `rem2rpx?: boolean | rem2rpxOptions`

rem 转 rpx 配置，默认 **不开启**，可传入配置项，配置项见 <https://www.npmjs.com/package/postcss-rem-to-responsive-pixel>
这个配置项代表插件内置了 `postcss-rem-to-responsive-pixel` ，不过默认不开启，传入一个 `true` 相当于传入配置:

```js
{
  // 32 意味着 1rem = 32rpx
  rootValue: 32,
  // 默认所有属性都转化
  propList: ['*'],
  // 转化的单位,可以变成 px / rpx
  transformUnit: 'rpx',
}
```

当然你也可以传入 `rem2rpxOptions` 这样一个 `object` 进行自定义

#### 为什么默认不开启？

1. 为了从 `2.x` 版本可以平滑的过渡到 `3.x`
2. 从我的视角看，内置 `postcss` 插件功能，虽然整体集成度上更高了，但是对其他开发者可能不是那么自由，比如在 `2.x` 时候，由于 `postcss-rem-to-responsive-pixel` 是外置的，所以开发者可以自由的决定它的加载顺序和加载逻辑，但是内置之后都是我决定的。不过内置好处也有，就是开箱即用

### 增强的配置项

- `cssChildCombinatorReplaceValue`, `cssSelectorReplacement.root`,`cssSelectorReplacement.universal` 现在都可以接受字符串数组了，它们可以自动展开，防止选择器格式化错误问题

### 修改的默认值

- `cssPreflightRange` 从 `'view'` 变为 `undefined`, 现在 `all` 的作用变成了在 `tailwindcss` 变量注入区域的选择器，添加一个 `:not(not)` 的选择器作为全局选择器的替代
- `cssSelectorReplacement.universal` 从 `'view'` 变为 `['view', 'text']`, 这意味着 `*` 选择器会被展开成 `view,text` 以及对应方式
- `cssChildCombinatorReplaceValue` 从 `'view + view'` 变为 `['view']`
- `replaceUniversalSelectorWith`,`jsEscapeStrategy`,`minifiedJs` 选项被删除，所以不再保留默认值

### 现在选项合并，数组默认行为变为覆盖，原先是合并

```js
const options = getOptions(input,defaults)
defaults: ['a','b'], input:['c'] 
// before: 
options == ['a','b','c']
// after:
options == ['c']
```

<https://github.com/sonofmagic/weapp-tailwindcss/issues/259>
