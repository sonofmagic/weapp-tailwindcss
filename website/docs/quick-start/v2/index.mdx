# Tailwindcss 2.x

目前，有些用户由于现有的项目，已经是 `webpack 4`, `postcss 7.x` 且无法往上升级，但是又想要使用 `tailwindcss`, 

所以写了这个文档作为参考，在现有版本的情况下，**不推荐**任何的新项目使用

## 安装

在这种条件下，只能使用 `tailwindcss 2.x` 版本。

参考 https://v2.tailwindcss.com/docs/installation#post-css-7-compatibility-build 中的安装方式

安装好之后，一定要打开 `jit` 模式, https://github.com/icebreaker-trash/uni-app-vue2-tailwind-hbuilder-template/blob/master/tailwind.config.js

具体更多的细节，详见下方模板代码。

## vue2 hbuilderx 参考模板

注意，一定要在开发环境中设置 

```js
process.env.TAILWIND_MODE = "watch"
```

才能正常热更新

模板源代码地址:

https://github.com/icebreaker-template/uni-app-vue2-tailwind-hbuilder-template


