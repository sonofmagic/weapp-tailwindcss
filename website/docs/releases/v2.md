# 2.x 版本新特性

这个版本新增了 `UnifiedWebpackPluginV5`
和 `UnifiedViteWeappTailwindcssPlugin` 这种 `Unified` 开头的插件。

它们能够自动识别并精确处理所有 `tailwindcss` 的工具类。这意味着它可以同时处理所有文件中的静态或动态的 `class`。

相比`v1`版本只有处理`wxss`,`wxml`静态`class`的能力，使用`v2`版本新的插件，你再也不需要在 `js` 里引入并调用标记方法 `replaceJs`了！`2.x` 插件有精准转化 `js`/`jsx` 的能力，大大提升了 `taro` 这种动态模板框架的开发体验。

## 新插件介绍

`UnifiedWebpackPluginV5` 是一个核心插件，所有使用 `webpack` 进行打包的框架都可以使用它，只需要传入 `appType` 配置项: `uni-app`/`taro`/`rax`/`remax`/`mpx` 等等，如果不传的话，插件会去猜测公共的样式文件位置，并进行转化(有可能不准确)。

目前，这个方案只支持 `tailwindcss v3.2.2` 以上版本和 `webpack5`。同时这个方案依赖 `monkey patch`，所以你应该把

```json
 "scripts": {
+  "postinstall": "weapp-tw patch"
 }
```

加入你的 `package.json`。当然在安装或者更新 `tailwindcss` 后，手动执行 `npx weapp-tw patch` 效果也是一样的，看到 `patch .... successfully` 表示成功。

`UnifiedViteWeappTailwindcssPlugin` 为 `vite` 专用插件，配置项和使用方式也是和上面一致的。
