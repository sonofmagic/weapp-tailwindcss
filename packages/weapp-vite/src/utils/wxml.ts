import { Parser } from 'htmlparser2'

const srcImportTagsMap: Record<string, string[]> = {
  audio: ['src', 'poster'],
  video: ['src', 'poster'],
  image: ['src'],
  // https://developers.weixin.qq.com/miniprogram/dev/reference/wxs/01wxs-module.html
  wxs: ['src'],
  // https://developers.weixin.qq.com/miniprogram/dev/reference/wxml/import.html
  import: ['src'],
  include: ['src'],
}

export interface WxmlDep {
  tagName: string
  start: number
  end: number
  quote: string | null | undefined
  name: string
  value: string
}

export function getDeps(wxml: string) {
  const deps: WxmlDep[] = []
  let currentTagName = ''
  let importAttrs: undefined | string[]
  const parser = new Parser({
    onopentagname(name) {
      currentTagName = name
      importAttrs = srcImportTagsMap[currentTagName]
    },
    onattribute(name, value, quote) {
      if (importAttrs) {
        for (const attrName of importAttrs) {
          if (attrName === name) {
            deps.push({
              name,
              value,
              quote,
              tagName: currentTagName,
              start: parser.startIndex,
              end: parser.endIndex,
            })
            return
          }
        }
      }
    },
    // ontext(data) {
    //   console.log('ontext', data)
    // },
    onclosetag() {
      currentTagName = ''
      importAttrs = undefined
    },
  })
  parser.write(
    wxml,
  )
  parser.end()
  return {
    deps,
  }
}
