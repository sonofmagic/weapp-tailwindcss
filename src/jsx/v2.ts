import { replaceWxml } from '@/wxml/shared'
import { parse, traverse, generate } from '@/babel'
import type { TraverseOptions, IJsxHandlerOptions } from '@/types'
import type { Node, Identifier, StringLiteral, Expression, PrivateName } from '@babel/types'

import { templeteHandler } from '@/wxml/utils'
import { defu } from '@/shared'
const StartMatchKeyMap: Record<'react' | 'vue2' | 'vue3' | string, string[]> = {
  react: ['className', 'hoverClass', 'hoverClassName', 'class', 'hover-class'],
  vue2: ['class', 'staticClass'], // 'hover-class' 在 'attrs' 里
  vue3: ['class', 'hover-class']
}

export function getKey (node: Identifier | StringLiteral | Expression | PrivateName): string {
  if (node.type === 'Identifier') {
    return node.name
  }
  if (node.type === 'StringLiteral') {
    return node.value
  }
  return ''
}

function isObjectKey (type: string) {
  return ['Identifier', 'StringLiteral'].includes(type)
}

export function jsxHandler (
  rawSource: string,
  opt: IJsxHandlerOptions = {
    framework: 'react'
  }
) {
  const { framework, escapeEntries } = opt
  const ast = parse(rawSource, {
    sourceType: 'unambiguous'
  }) as Node
  const matchKeys = StartMatchKeyMap[framework as string] ?? StartMatchKeyMap.react
  const isVue2 = framework === 'vue2'
  const isVue3 = framework === 'vue3'
  // https://astexplorer.net/
  let startFlag = false

  const options: TraverseOptions<Node> = {
    ObjectProperty: {
      enter (path) {
        if (isObjectKey(path.node.key.type)) {
          const keyStr = getKey(path.node.key)
          if (matchKeys.includes(keyStr)) {
            startFlag = true
          }
          if (isVue2 && keyStr === 'attrs' && path.node.value.type === 'ObjectExpression') {
            const idx = path.node.value.properties.findIndex((x) => {
              return x.type === 'ObjectProperty' && isObjectKey(x.key.type) && getKey(x.key) === 'hover-class'
            })
            const hoverClassNode = path.node.value.properties[idx]
            if (idx > -1 && hoverClassNode.type === 'ObjectProperty') {
              if (hoverClassNode.value.type === 'StringLiteral') {
                hoverClassNode.value.value = replaceWxml(hoverClassNode.value.value, {
                  keepEOL: true,
                  escapeEntries
                })
              }
            }
          }
        }
      },
      exit (path) {
        if (['Identifier', 'StringLiteral'].includes(path.node.key.type) && matchKeys.includes(getKey(path.node.key))) {
          startFlag = false
        }
      }
    },
    StringLiteral: {
      enter (path) {
        if (startFlag) {
          path.node.value = replaceWxml(path.node.value, {
            keepEOL: true,
            escapeEntries
          })
        }
      }
    },
    noScope: true
  }
  if (isVue3) {
    options.CallExpression = {
      enter (path) {
        const hit = path.node.arguments[0]
        if (hit && hit.type === 'StringLiteral') {
          hit.value = templeteHandler(hit.value, {
            keepEOL: true
          })
        }
      }
    }
  }
  // const vue3StaticVNodeStartFlag = false
  traverse(ast, options)
  // @ts-ignore
  return generate(ast)
}

export function createJsxHandler (options: IJsxHandlerOptions) {
  return (rawSource: string, opt?: IJsxHandlerOptions) => {
    return jsxHandler(rawSource, defu(opt, options))
  }
}
