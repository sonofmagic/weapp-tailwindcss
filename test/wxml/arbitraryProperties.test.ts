import { templateHandler } from '#test/v2/wxml'

describe('arbitrary properties', () => {
  test('[mask-type:luminance]', () => {
    expect(
      templateHandler(`<div class="[mask-type:luminance]">
      <!-- ... -->
    </div>`)
    ).toMatchSnapshot()
  })

  test('[mask-type:luminance] hover:[mask-type:alpha]', () => {
    expect(
      templateHandler(`<div class="[mask-type:luminance] hover:[mask-type:alpha]">
      <!-- ... -->
    </div>`)
    ).toMatchSnapshot()
  })

  test('[--scroll-offset:56px] lg:[--scroll-offset:44px]', () => {
    expect(
      templateHandler(`<div class="[--scroll-offset:56px] lg:[--scroll-offset:44px]">
      <!-- ... -->
    </div>`)
    ).toMatchSnapshot()
  })
})
