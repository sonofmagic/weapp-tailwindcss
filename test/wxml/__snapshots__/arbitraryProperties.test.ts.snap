// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`arbitrary properties > [--scroll-offset:56px] lg:[--scroll-offset:44px] 1`] = `
"<div class=\\"_--scroll-offsetc56px_ lgc_--scroll-offsetc44px_\\">
      <!-- ... -->
    </div>"
`;

exports[`arbitrary properties > [mask-type:luminance] 1`] = `
"<div class=\\"_mask-typecluminance_\\">
      <!-- ... -->
    </div>"
`;

exports[`arbitrary properties > [mask-type:luminance] hover:[mask-type:alpha] 1`] = `
"<div class=\\"_mask-typecluminance_ hoverc_mask-typecalpha_\\">
      <!-- ... -->
    </div>"
`;
